File: src/RPC.php
<?php

declare(strict_types=1);

namespace Spiral\Goridge;

final class RPC
{
    /**
     * @var Relay
     */
    private $relay;

    /**
     * @var int
     */
    private $seq;

    /**
     * @param Relay $relay
     */
    public function __construct(Relay $relay)
    {
        $this->relay = $relay;
    }

    /**
     * @param string $uri
     * @param array  $options
     *
     * @return self
     */
    public static function connect(string $uri, array $options = []): self
    {
        return new self(Relay::connect($uri, $options));
    }

    /**
     * @param string $method
     * @param mixed  $payload
     * @param bool   $raw
     *
     * @return mixed
     */
    public function call(string $method, $payload, bool $raw = false)
    {
        $call = goridge_pack($method . \pack("P", $this->seq), Frame::PAYLOAD_CONTROL | Frame::PAYLOAD_RAW);

        if ($raw) {
            $data = goridge_pack($payload, Frame::PAYLOAD_RAW);
        } else {
            $body = \json_encode($payload);

            if (\JSON_ERROR_NONE !== \json_last_error()) {
                throw Exception\JSONException::last();
            }

            $data = goridge_pack($body);
        }

        $this->relay->send($call . $data);

        $control = $this->relay->receive();

        if (!$control->isControl()) {
            throw Exception\RPCException::noControl();
        }

        $rpc = \unpack("Ps", \substr($control->body, -8));
        $rpc['m'] = \substr($control->body, 0, -8);

        if ($rpc["m"] !== $method || $rpc['s'] != $this->seq) {
            throw Exception\RPCException::controlMismatch();
        }

        $this->seq++;

        $response = $this->relay->receive();

        if ($response->isError()) {
            throw Exception\ServiceException::forResponse($response);
        }

        return $response->isRaw() ? $response->body : \json_decode($response->body, true);
    }
}

File: ext/src/RPC.cpp
#include "RPC.hpp"
#include "Exceptions.hpp"

namespace Goridge {
    RPC::RPC(Relay * relay): relay(relay) {
    }

    Frame * RPC::call(const char * method, const char * payload, size_t size, bool raw) {
        size_t method_size = strlen(method);
        size_t request_size = method_size + 8;

        size_t l_method = HEADER_SIZE + request_size;
        size_t l_payload = HEADER_SIZE + size;

        char request[request_size];
        char binary[l_method + l_payload];

        memcpy(request, method, method_size);
        memcpy(request + strlen(method), m_seq.chars, 8);

        Frame::pack(binary, request, request_size, PAYLOAD_CONTROL | PAYLOAD_RAW);
        Frame::pack(binary + l_method, payload, size, raw ? PAYLOAD_RAW : 0);

        relay->send(binary, l_method + l_payload);
        Frame * control = relay->receive();

        if (!control->isControl()) {
            throw RPCException("RPC response header is missing");
        }

        uint64_t _seq;
        char _method[control->size - 7];

        memcpy(&_seq, control->body + control->size - 8, 8);
        memcpy(&_method, control->body, control->size - 8);

        _method[control->size - 8] = '\0';

        if (strcmp(method, _method) != 0) {
            throw RPCException("RPC method mismatch");
        }

        if (m_seq.integer != _seq) {
            throw RPCException("RPC seq mismatch");
        }

        m_seq.integer++;

        return relay->receive();
    }

    RPC::~RPC() {
    }
}  // namespace Goridge
